presto-main/src/main/java/com/facebook/presto/cost/PlanNodeCostEstimate.java

```diff
@@ -0,0 +1,153 @@
+    public static final PlanNodeCostEstimate ZERO_COST = PlanNodeCostEstimate.builder().build();
+
+    private final double cpuCost;
+    private final double memoryCost;
+    private final double networkCost;
+
+    private PlanNodeCostEstimate(double cpuCost, double memoryCost, double networkCost)
+    {
+        this.cpuCost = cpuCost;
+        this.memoryCost = memoryCost;
+        this.networkCost = networkCost;
+    }
+
+    public double getCpuCost()
+    {
+        return cpuCost;
+    }
+
+    public double getMemoryCost()
+    {
+        return memoryCost;
+    }
+
+    public double getNetworkCost()
+    {
+        return networkCost;
+    }
+
+    public boolean isUnknown()
+    {
+        return isNaN(cpuCost) || isNaN(memoryCost) || isNaN(networkCost);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "PlanNodeCostEstimate{" +
+                "cpuCost=" + cpuCost +
+                ", memoryCost=" + memoryCost +
+                ", networkCost=" + networkCost + '}';
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        PlanNodeCostEstimate that = (PlanNodeCostEstimate) o;
+        return Objects.equals(cpuCost, that.cpuCost) &&
+                Objects.equals(memoryCost, that.memoryCost) &&
+                Objects.equals(networkCost, that.networkCost);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(cpuCost, memoryCost, networkCost);
+    }
+
+    public PlanNodeCostEstimate add(PlanNodeCostEstimate other)
+    {
+        return new PlanNodeCostEstimate(
+                cpuCost + other.cpuCost,
+                memoryCost + other.memoryCost,
+                networkCost + other.networkCost);
+    }
+
+    public static PlanNodeCostEstimate networkCost(double networkCost)
+    {
+        return builder().setNetworkCost(networkCost).build();
+    }
+
+    public static PlanNodeCostEstimate cpuCost(double cpuCost)
+    {
+        return builder().setCpuCost(cpuCost).build();
+    }
+
+    public static PlanNodeCostEstimate memoryCost(double memoryCost)
+    {
+        return builder().setCpuCost(memoryCost).build();
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static final class Builder
+    {
+        private double cpuCost = 0;
+        private double memoryCost = 0;
+        private double networkCost = 0;
+
+        public Builder setFrom(PlanNodeCostEstimate otherStatistics)
+        {
+            return setCpuCost(otherStatistics.getCpuCost())
+                    .setMemoryCost(otherStatistics.getMemoryCost())
+                    .setNetworkCost(otherStatistics.getNetworkCost());
+        }
+
+        public Builder setCpuCost(double cpuCost)
+        {
+            this.cpuCost = cpuCost;
+            return this;
```

Comment:
IMO when fluent builder wants to guarantee it returns `this` it should advertise this (e.g. https://github.com/prestodb/presto/pull/8294). Unless I'm the only person who thinks making builder class immutable is not totally absurd idea.